### ISSUE TITLE
Can't Trade with Myself in Pokemon Red/Bluw

### ISSUE THREAD

**Issue by Myusicman** [visualboyadvance-m/visualboyadvance-m#604]
### Expected behavior
I want to trade with myself in Pokemon Red/Blue using two screens within one computer

### Actual behavior
The screens do not register the cable link

### Steps to reproduce the behavior
Click on the VBA icon twice to start up two separate screens
For each one, go to Link -> Type -> Cable
For Screen 1, go to Link -> Start Network Link -> Server -> 2 Players -> Start
For Screen 2, go to Link -> Start Network Link -> Client -> Start
Start up games for both screens and open desired save files
At a Pokemon Center, go to the far right in each screen and watch as the lady there refuses to let you start a trade.

### ROM details

Which games are affected by the issue (please state if GBA or GB game):
Pokemon - Red Version (UE)
Pokemon - Blue Version (UE)

Include screenshots of ROM Information (File > ROM Information):

### Option details

Video driver (simple, opengl):

Sound driver (directsound, xaudio, openal, SDL):

### Build details

Operating System (Windows, Mac, Linux (state distribution), etc.): Windows 10

If using Linux, specify if using xorg or Wayland:

Version of code (pre-built binary version, or commit ref, or just "master"):

### Hardware details:

If your joystick or gamepad isn't working in Linux, please try to troubleshoot
using the information on this page:
https://wiki.archlinux.org/index.php/Gamepad **BEFORE** filing an issue. We use
SDL for joystick support, any problems will generally be an issue with SDL and
**NOT** the emulator.

If relevant, the joystick you are using and the drivers you are using for it:

//Let me know if you need more info.  Thank you for your time.  Also, apologies to sending this to the wrong forum the first time.

**Comment by ZachBacon** [comment]
@Myusicman do you mine testing this https://github.com/visualboyadvance-m/visualboyadvance-m/issues/354#issuecomment-58216278 This contains some fixes for link connections.

**Comment by Myusicman** [comment]
@ZachBacon I downloaded the 64-bit files, but still no luck.  The link setup seems to work, it's just the games don't recognize a link active.

I feel like it's worth noting that in my older laptop, the link actually did go through the game, but whenever I was given the option to either trade or fight, one screen would freeze up entirely (my laptop was over 5 years old and had only 4GB Ram, which caused the games to lag whenever more than one screen was active).  Only one time was I able to get both screens into the trading area, but later on one of the screens got so glitched that it was impossible for me to select a Pokemon to trade.

Any other way I can resolve this?

**Comment by rkitover** [comment]
Are these gameboy advance or gameboy (color) roms?

**Comment by Myusicman** [comment]
@rkitover Gameboy roms.  Does it matter if they're unzipped or not?

**Comment by ZachBacon** [comment]
@Myusicman it doesn't matter, just with Pokemon games some are re-releases in the gba like fire-red.

**Comment by Myusicman** [comment]
Alright.  Well it is gen 1, so it's the original Red and Blue, so it would be a GB rom, not Advanced

**Comment by rkitover** [comment]
This is consistent with the link behavior @denisfa described for GB roms.

**Comment by Myusicman** [comment]
@rkitover I see.  Has he come up with a solution for this yet?

**Comment by denisfa** [comment]
@Myusicman Sorry, this is a known issue for a while. We will eventually tackle it.. The behaviour I noticed is a little different from your description. My config is:

- Link -> Type -> Game Boy
- Link -> Configure -> 120000 (timeout value)

Make sure your emulated system is set to Game Boy. You will very likely see glimpses of connection, although it will(should) be unusable due to desync (you can test multiple timeout values here). If you have time, you can confirm this yourself.

**Comment by Myusicman** [comment]
@denisfa The GameBoy link works, however, as you mentioned, the desync does make it impossible to use.  While the game registers the link, it has the same problem as my old computer did where one of the screens eventually freezes up.

As you find a way to resolve this, I will find ways to pass the time.  Maybe I can finish up Blue and Yellow or just start a Gen 2 game.  Thank you for your help! :)

**Comment by rkitover** [comment]
If the link feature is important for you, you can try doing a GBA pokemon game, GBA link is far more solid.

**Comment by Myusicman** [comment]
@ZachBacon @rkitover @denisfa Hello, it has been about two weeks since I last replied.  I hope you are all doing well!  Any update on the GameBoy/Cable Link issue?

**Comment by rkitover** [comment]
@Myusicman I'm afraid it will take us much longer than this to get to this issue, you are being way too optimistic.

**Comment by Myusicman** [comment]
@rkitover Ah, my apologies.  I have no knowledge on computer programming and coding, but I do know that it can be a long and difficult process.  Hopefully you guys will be able to resolve the issue!  As I said earlier, I will be waiting patiently in the meantime :)

Thanks again for listening, it means a lot that you guys are willing to put effort into things like this.  I hope you all have a great day!

**Comment by rkitover** [comment]
We have actually been discussing link in detail over the past few weeks, and we've come to the conclusion that for GB link, the main issues are synchronization.

So this is definitely on our radar as one of the most important things to fix.

**Comment by Myusicman** [comment]
@rkitover Ah I see, good to know, then!

**Comment by Squall-Leonhart** [comment]
> We have actually been discussing link in detail over the past few weeks, and we've come to the conclusion that for GB link, the main issues are synchronization.

pipelining is the only way to do it right,  trying to do it via tcpip sockets just isn't going to cut it, it.

**Comment by Myusicman** [comment]
Hey guys!  I know it's been a while, but is there any update on the fixes for the emulator?  If so, lemme know!

**Comment by rkitover** [comment]
No, unfortunately we have not started on this issue yet.

**Comment by Myusicman** [comment]
Really?  I imagine that you all must have been very busy over these past few months, then.  Do you know when you will be able to start working on this issue?

**Comment by rkitover** [comment]
Well you can see what we're doing in the commit log:

https://github.com/visualboyadvance-m/visualboyadvance-m/commits/master

the other primary developer on this project, denisfa, just left. So it's just me now.

My primary goal is fixing any major usability issues and getting the next release out.

I may have more time soon to work on this project, in which case link issues will be looked at for sure.

**Comment by nklein1** [comment]
I'd like to add a +1 to this task. Thanks for your hard work!

**Comment by GregTheHun** [comment]
Yeah, thank you for the hard work, +1 this too, I'd really love to have my nostalgia fed with all 151 Pokemon from Red/Blue.

:-)

**Comment by MaxLempira** [comment]
Hello, any news about this? I think this is an important issue. Also... I don't know if this could help but, TGB Dual does this without any problem... maybe a check on the code could help?

https://github.com/libertyernie/tgbdual_L

Just trying to help

**Comment by Karlings** [comment]
Yeah that emulator is the only way to trade with the first and second gen...

**Comment by kevinpirola** [comment]
is there any workaround yet to this issue? I'm committed in a 50+ hours game and I would like to have trading usable...
is there a way to open the save in any other emulator that have the link working?

**Comment by Walter135** [comment]
Guys you can use bgbw64 emulator, this has a good trade connections between pokemon firrst and second generations

**Comment by WhitleyStriber** [comment]
Would love to get this working.

**Comment by sirj0k3r** [comment]
In case there's anyone keeping an eye on this, I also noticed the same issue on macOS.

Game Boy Advance games work fine (as long as Cable is selected).
Game Boy and Game Boy Color games do not seem to connect (neither via Cable nor Game Boy).
